// breezy-core/src/security/migration/static-models.js
//*****************STATIC MODELS
'use strict'

let sys_admin_managerModel = {
  userId: 'USER_ID',
  companyId: 'COMPANY_ID',
  teamIds: [],
  roles: ['sys_admin', 'manager'],
  positionRoles: {manager: ['POSITION_ID']},
  permissions:
    {
      admin: 15,
      hris: 15,
      team: 15,
      sso: 15,
      gdpr: 15,
      eeoc: 6,
      tag: 15,
      task: 15,
      smsEmail: 15,
      integration: 15,
      customField: 15,
      company: 15,
      user: 15,
      talentSearch: 2,
      selfTask: 15,
      position: 15,
      userAutoJoin: 15,
      approval: 15,
      socialDiscovery: 6,
      mergeCandidate: 6,
      careerPortal: 6,
      recruiterPortal: 6,
      helloMessenger: 6,
      template: 15,
      campaign: 15,
      pipeline: 15,
      questionnaire: 15,
      interviewGuide: 15,
      scorecard: 15,
      schedulingLink: 15,
      assessment: 15,
      backgroundCheck: 15,
      security: 15,
      report: 15,
      manager:
        {
          candidate: 15,
          resume: 3,
          experience: 15,
          meeting: 15,
          reference: 2,
          document: 11,
          detail: 15,
          discussionMessage: 15,
          discussionActivity: 2,
          teamScorecard: 2,
          requestScorecard: 3,
          reminder: 11,
          gdprConsent: 3,
          assignCandidate: 6,
          gdprData: 3,
          stage: 6,
          candidateSource: 15,
          candidateContact: 15,
          sharedNotes: 2,
          notes: 15,
          offer: 3,
          candidateScore: 15,
          candidateScorecard: 15,
          candidateQuestionnaire: 15,
          candidateCopy: 15,
          candidateDelete: 15,
          candidateMove: 15,
          candidateStatus: 15
        }
    }
}

let sys_admin_position_memberModel = {
  userId: 'USER_ID',
  companyId: 'COMPANY_ID',
  teamIds: [],
  roles: ['sys_admin', 'position_member'],
  positionRoles: {position_member: ['POSITION_ID']},
  permissions:
    {
      admin: 15,
      hris: 15,
      team: 15,
      sso: 15,
      gdpr: 15,
      eeoc: 6,
      tag: 15,
      task: 15,
      smsEmail: 15,
      integration: 15,
      customField: 15,
      company: 15,
      user: 15,
      talentSearch: 2,
      selfTask: 15,
      position: 15,
      userAutoJoin: 15,
      approval: 15,
      socialDiscovery: 6,
      mergeCandidate: 6,
      careerPortal: 6,
      recruiterPortal: 6,
      helloMessenger: 6,
      template: 15,
      campaign: 15,
      pipeline: 15,
      questionnaire: 15,
      interviewGuide: 15,
      scorecard: 15,
      schedulingLink: 15,
      assessment: 15,
      backgroundCheck: 15,
      security: 15,
      report: 15,
      position_member:
        {
          resume: 2,
          experience: 2,
          discussionMessage: 15,
          notes: 15,
          candidate: 2,
          candidateScorecard: 14,
          candidateQuestionnaire: 2,
          reminder: 11
        }
    }
}

let sys_admin_AllModel = {
  userId: 'USER_ID',
  companyId: 'COMPANY_ID',
  teamIds: [],
  roles: ['sys_admin', 'manager', 'position_member'],
  positionRoles:
    {
      position_member: ['POSITION_ID'],
      manager: ['POSITION_ID']
    },
  permissions:
    {
      admin: 15,
      hris: 15,
      team: 15,
      sso: 15,
      gdpr: 15,
      eeoc: 6,
      tag: 15,
      task: 15,
      smsEmail: 15,
      integration: 15,
      customField: 15,
      company: 15,
      user: 15,
      talentSearch: 2,
      selfTask: 15,
      position: 15,
      userAutoJoin: 15,
      approval: 15,
      socialDiscovery: 6,
      mergeCandidate: 6,
      careerPortal: 6,
      recruiterPortal: 6,
      helloMessenger: 6,
      template: 15,
      campaign: 15,
      pipeline: 15,
      questionnaire: 15,
      interviewGuide: 15,
      scorecard: 15,
      schedulingLink: 15,
      assessment: 15,
      backgroundCheck: 15,
      security: 15,
      report: 15,
      manager:
        {
          candidate: 15,
          resume: 3,
          experience: 15,
          meeting: 15,
          reference: 2,
          document: 11,
          detail: 15,
          discussionMessage: 15,
          discussionActivity: 2,
          teamScorecard: 2,
          requestScorecard: 3,
          reminder: 11,
          gdprConsent: 3,
          assignCandidate: 6,
          gdprData: 3,
          stage: 6,
          candidateSource: 15,
          candidateContact: 15,
          sharedNotes: 2,
          notes: 15,
          offer: 3,
          candidateScore: 15,
          candidateScorecard: 15,
          candidateQuestionnaire: 15,
          candidateCopy: 15,
          candidateDelete: 15,
          candidateMove: 15,
          candidateStatus: 15
        },
      position_member:
        {
          resume: 2,
          experience: 2,
          discussionMessage: 15,
          notes: 15,
          candidate: 2,
          candidateScorecard: 14,
          candidateQuestionnaire: 2,
          reminder: 11
        }
    }
}

let sys_adminModel = {
  userId: 'USER_ID',
  companyId: 'COMPANY_ID',
  teamIds: [],
  roles: ['sys_admin'],
  positionRoles: {},
  permissions:
    {
      admin: 15,
      hris: 15,
      team: 15,
      sso: 15,
      gdpr: 15,
      eeoc: 6,
      tag: 15,
      task: 15,
      smsEmail: 15,
      integration: 15,
      customField: 15,
      company: 15,
      user: 15,
      talentSearch: 2,
      selfTask: 15,
      position: 15,
      userAutoJoin: 15,
      approval: 15,
      socialDiscovery: 6,
      mergeCandidate: 6,
      careerPortal: 6,
      recruiterPortal: 6,
      helloMessenger: 6,
      template: 15,
      campaign: 15,
      pipeline: 15,
      questionnaire: 15,
      interviewGuide: 15,
      scorecard: 15,
      schedulingLink: 15,
      assessment: 15,
      backgroundCheck: 15,
      security: 15,
      report: 15
    }
}

let admin_managerModel = {
  userId: 'USER_ID',
  companyId: 'COMPANY_ID',
  teamIds: [],
  roles: ['admin', 'manager'],
  positionRoles: {manager: ['POSITION_ID']},
  permissions:
    {
      admin: 15,
      hris: 15,
      candidate: 15,
      userAutoJoin: 15,
      team: 15,
      approval: 15,
      sso: 15,
      socialDiscovery: 6,
      mergeCandidate: 6,
      gdpr: 15,
      eeoc: 6,
      careerPortal: 6,
      recruiterPortal: 6,
      helloMessenger: 6,
      interviewGuide: 15,
      questionnaire: 15,
      scorecard: 15,
      assessment: 15,
      backgroundCheck: 15,
      tag: 15,
      task: 15,
      integration: 15,
      position: 15,
      customField: 15,
      schedulingLink: 15,
      pipeline: 15,
      company: 15,
      template: 15,
      security: 15,
      campaign: 15,
      user: 15,
      resume: 15,
      experience: 15,
      selfTask: 15,
      smsEmail: 15,
      meeting: 15,
      reference: 15,
      document: 15,
      detail: 15,
      discussionMessage: 15,
      discussionActivity: 15,
      teamScorecard: 15,
      requestScorecard: 15,
      reminder: 15,
      gdprConsent: 15,
      gdprData: 15,
      stage: 15,
      candidateSource: 15,
      candidateContact: 15,
      sharedNotes: 15,
      notes: 15,
      report: 15,
      disposition: 15,
      offerTemplate: 15,
      assignCandidate: 15,
      offer: 15,
      talentSearch: 2,
      candidateScore: 15,
      candidateScorecard: 15,
      candidateQuestionnaire: 15,
      candidateCopy: 15,
      candidateDelete: 15,
      candidateStatus: 15,
      candidateMove: 15,
      manager: {}
    }
}

let admin_position_memberModel = {
  userId: 'USER_ID',
  companyId: 'COMPANY_ID',
  teamIds: [],
  roles: ['admin', 'position_member'],
  positionRoles: {position_member: ['POSITION_ID']},
  permissions:
    {
      admin: 15,
      hris: 15,
      candidate: 15,
      userAutoJoin: 15,
      team: 15,
      approval: 15,
      sso: 15,
      socialDiscovery: 6,
      mergeCandidate: 6,
      gdpr: 15,
      eeoc: 6,
      careerPortal: 6,
      recruiterPortal: 6,
      helloMessenger: 6,
      interviewGuide: 15,
      questionnaire: 15,
      scorecard: 15,
      assessment: 15,
      backgroundCheck: 15,
      tag: 15,
      task: 15,
      integration: 15,
      position: 15,
      customField: 15,
      schedulingLink: 15,
      pipeline: 15,
      company: 15,
      template: 15,
      security: 15,
      campaign: 15,
      user: 15,
      resume: 15,
      experience: 15,
      selfTask: 15,
      smsEmail: 15,
      meeting: 15,
      reference: 15,
      document: 15,
      detail: 15,
      discussionMessage: 15,
      discussionActivity: 15,
      teamScorecard: 15,
      requestScorecard: 15,
      reminder: 15,
      gdprConsent: 15,
      gdprData: 15,
      stage: 15,
      candidateSource: 15,
      candidateContact: 15,
      sharedNotes: 15,
      notes: 15,
      report: 15,
      disposition: 15,
      offerTemplate: 15,
      assignCandidate: 15,
      offer: 15,
      talentSearch: 2,
      candidateScore: 15,
      candidateScorecard: 15,
      candidateQuestionnaire: 15,
      candidateCopy: 15,
      candidateDelete: 15,
      candidateStatus: 15,
      candidateMove: 15,
      position_member: {}
    }
}

let admin_AllModel = {
  userId: 'USER_ID',
  companyId: 'COMPANY_ID',
  teamIds: [],
  roles: ['admin', 'manager', 'position_member'],
  positionRoles:
    {
      position_member: ['POSITION_ID'],
      manager: ['POSITION_ID']
    },
  permissions:
    {
      admin: 15,
      hris: 15,
      candidate: 15,
      userAutoJoin: 15,
      team: 15,
      approval: 15,
      sso: 15,
      socialDiscovery: 6,
      mergeCandidate: 6,
      gdpr: 15,
      eeoc: 6,
      careerPortal: 6,
      recruiterPortal: 6,
      helloMessenger: 6,
      interviewGuide: 15,
      questionnaire: 15,
      scorecard: 15,
      assessment: 15,
      backgroundCheck: 15,
      tag: 15,
      task: 15,
      integration: 15,
      position: 15,
      customField: 15,
      schedulingLink: 15,
      pipeline: 15,
      company: 15,
      template: 15,
      security: 15,
      campaign: 15,
      user: 15,
      resume: 15,
      experience: 15,
      selfTask: 15,
      smsEmail: 15,
      meeting: 15,
      reference: 15,
      document: 15,
      detail: 15,
      discussionMessage: 15,
      discussionActivity: 15,
      teamScorecard: 15,
      requestScorecard: 15,
      reminder: 15,
      gdprConsent: 15,
      gdprData: 15,
      stage: 15,
      candidateSource: 15,
      candidateContact: 15,
      sharedNotes: 15,
      notes: 15,
      report: 15,
      disposition: 15,
      offerTemplate: 15,
      assignCandidate: 15,
      offer: 15,
      talentSearch: 2,
      candidateScore: 15,
      candidateScorecard: 15,
      candidateQuestionnaire: 15,
      candidateCopy: 15,
      candidateDelete: 15,
      candidateStatus: 15,
      candidateMove: 15,
      manager: {},
      position_member: {}
    }
}

let adminModel = {
  userId: 'USER_ID',
  companyId: 'COMPANY_ID',
  teamIds: [],
  roles: ['admin'],
  positionRoles: {},
  permissions:
    {
      admin: 15,
      hris: 15,
      candidate: 15,
      userAutoJoin: 15,
      team: 15,
      approval: 15,
      sso: 15,
      socialDiscovery: 6,
      mergeCandidate: 6,
      gdpr: 15,
      eeoc: 6,
      careerPortal: 6,
      recruiterPortal: 6,
      helloMessenger: 6,
      interviewGuide: 15,
      questionnaire: 15,
      scorecard: 15,
      assessment: 15,
      backgroundCheck: 15,
      tag: 15,
      task: 15,
      integration: 15,
      position: 15,
      customField: 15,
      schedulingLink: 15,
      pipeline: 15,
      company: 15,
      template: 15,
      security: 15,
      campaign: 15,
      user: 15,
      resume: 15,
      experience: 15,
      selfTask: 15,
      smsEmail: 15,
      meeting: 15,
      reference: 15,
      document: 15,
      detail: 15,
      discussionMessage: 15,
      discussionActivity: 15,
      teamScorecard: 15,
      requestScorecard: 15,
      reminder: 15,
      gdprConsent: 15,
      gdprData: 15,
      stage: 15,
      candidateSource: 15,
      candidateContact: 15,
      sharedNotes: 15,
      notes: 15,
      report: 15,
      disposition: 15,
      offerTemplate: 15,
      assignCandidate: 15,
      offer: 15,
      talentSearch: 2,
      candidateScore: 15,
      candidateScorecard: 15,
      candidateQuestionnaire: 15,
      candidateCopy: 15,
      candidateDelete: 15,
      candidateStatus: 15,
      candidateMove: 15
    }
}

let member_managerModel = {
  userId: 'USER_ID',
  companyId: 'COMPANY_ID',
  teamIds: [],
  roles: ['member', 'manager'],
  positionRoles: {manager: ['POSITION_ID']},
  permissions:
    {
      position: 3,
      task: 6,
      selfTask: 15,
      talentSearch: 2,
      manager:
        {
          hris: 3,
          position: 6,
          candidate: 15,
          assessment: 3,
          backgroundCheck: 3,
          tag: 15,
          task: 15,
          customField: 15,
          resume: 3,
          experience: 15,
          smsEmail: 3,
          meeting: 15,
          reference: 2,
          document: 11,
          detail: 15,
          discussionMessage: 15,
          discussionActivity: 2,
          teamScorecard: 2,
          requestScorecard: 3,
          reminder: 11,
          gdprConsent: 3,
          assignCandidate: 6,
          gdprData: 3,
          stage: 6,
          candidateSource: 15,
          candidateContact: 15,
          sharedNotes: 2,
          notes: 15,
          report: 2,
          offer: 3,
          candidateScore: 15,
          candidateScorecard: 15,
          questionnaire: 2,
          candidateQuestionnaire: 15,
          candidateCopy: 15,
          candidateDelete: 15,
          candidateMove: 15,
          candidateStatus: 15
        }
    }
}

let member_position_memberModel = {
  userId: 'USER_ID',
  companyId: 'COMPANY_ID',
  teamIds: [],
  roles: ['member', 'position_member'],
  positionRoles: {position_member: ['POSITION_ID']},
  permissions:
    {
      position: 3,
      task: 6,
      selfTask: 15,
      talentSearch: 2,
      position_member:
        {
          resume: 2,
          experience: 2,
          discussionMessage: 15,
          notes: 15,
          candidate: 2,
          candidateScorecard: 14,
          candidateQuestionnaire: 2,
          reminder: 11
        }
    }
}

let member_AllModel = {
  userId: 'USER_ID',
  companyId: 'COMPANY_ID',
  teamIds: [],
  roles: ['member', 'manager', 'position_member'],
  positionRoles:
    {
      position_member: ['POSITION_ID'],
      manager: ['POSITION_ID']
    },
  permissions:
    {
      position: 3,
      task: 6,
      selfTask: 15,
      talentSearch: 2,
      manager:
        {
          hris: 3,
          position: 6,
          candidate: 15,
          assessment: 3,
          backgroundCheck: 3,
          tag: 15,
          task: 15,
          customField: 15,
          resume: 3,
          experience: 15,
          smsEmail: 3,
          meeting: 15,
          reference: 2,
          document: 11,
          detail: 15,
          discussionMessage: 15,
          discussionActivity: 2,
          teamScorecard: 2,
          requestScorecard: 3,
          reminder: 11,
          gdprConsent: 3,
          assignCandidate: 6,
          gdprData: 3,
          stage: 6,
          candidateSource: 15,
          candidateContact: 15,
          sharedNotes: 2,
          notes: 15,
          report: 2,
          offer: 3,
          candidateScore: 15,
          candidateScorecard: 15,
          questionnaire: 2,
          candidateQuestionnaire: 15,
          candidateCopy: 15,
          candidateDelete: 15,
          candidateMove: 15,
          candidateStatus: 15
        },
      position_member:
        {
          resume: 2,
          experience: 2,
          discussionMessage: 15,
          notes: 15,
          candidate: 2,
          candidateScorecard: 14,
          candidateQuestionnaire: 2,
          reminder: 11
        }
    }
}

let memberModel = {
  userId: 'USER_ID',
  companyId: 'COMPANY_ID',
  teamIds: [],
  roles: ['member'],
  positionRoles: {},
  permissions: {position: 3, task: 6, selfTask: 15, talentSearch: 2}
}

let sys_admin = {
  all: sys_admin_AllModel,
  manager: sys_admin_managerModel,
  member: sys_admin_position_memberModel,
  none: sys_adminModel
}

let admin = {
  all: admin_AllModel,
  manager: admin_managerModel,
  member: admin_position_memberModel,
  none: adminModel
}

let member = {
  all: member_AllModel,
  none: memberModel,
  member: member_position_memberModel,
  manager: member_managerModel
}

module.exports = {
  sys_admin,
  admin,
  member
}

//*****************
